lib/Geo/ShapeContainer.cpp

```diff
@@ -676,20 +673,23 @@ bool ShapeContainer::intersects(S2Polyline const* other) const {
 }
 
 namespace {
+
 bool intersectRectPolygon(S2LatLngRect const* rect, S2Polygon const* poly) {
   // only used in legacy situations
-  if (rect->is_full()) {
-    return true;  // rectangle spans entire sphere
-  } else if (rect->is_point()) {
-    return poly->Contains(rect->lo().ToPoint());  // easy case
-  } else if (!rect->Intersects(poly->GetRectBound())) {
-    return false;  // cheap rejection
+  if (ADB_UNLIKELY(rect->is_point())) {
+    return poly->Contains(rect->lo().ToPoint());
+  }
+  auto bound = poly->GetRectBound();
+  if (ADB_LIKELY(!rect->Intersects(bound))) {
+    return false;
+  } else if (ADB_LIKELY(rect->Contains(bound))) {
+    return true;
   }
   auto rectPoly = ::latLngRectToPolygon(rect);
-  return poly->Intersects(&rectPoly);
+  return S2BooleanOperation::Intersects(poly->index(), rectPoly.index());
```

Comment:
This function called in Intersects, we use it because it public and `Intersects` duplicate our checks